[Feat] 회원가입 기능 구현

prozen99 opened the pull request:
# ⭐️ Pull Request 요약

- 전통적 방식의 loginid 와 password를 통한 회원가입 구현 

## 🛠️ 변경 사항

- 이 PR의 변경 사항

## 💡 관련 이슈

- #5

## ⏱️ 작업 기간

- 작업 시작일: (2024-08-29)
- 작업 종료일: (2024-08-29)

## 📌 유의사항

-

## 🔗 참고문헌

- 서적

prozen99 committed:
feat :파일 추가
회원가입을 위해서
MemberController
MemberRepository
MemberService 구성

prozen99 committed:
Refactor: yaml 파일
local db로 수정

Galmaeki commented on src/main/java/discordstudy/calender/domain/member/controller/MemberController.java:
```diff
@@ -0,0 +1,26 @@
+    private final MemberService memberService;
+
+    @PostMapping("/signup")
+    public ResponseEntity<Member> signup(@RequestBody Member member )
```
DB에 들어가는 엔티티를 그대로 리턴하는것은 위험
요청도 객체를 그대로 받는것은 어색함
Request, Response Dto 만드셈

Galmaeki commented on src/main/java/discordstudy/calender/domain/member/controller/MemberController.java:
```diff
@@ -0,0 +1,26 @@
+    public ResponseEntity<Member> signup(@RequestBody Member member )
+    {
+        Member savedmember = memberService.registermember(member);
+        return ResponseEntity.ok(savedmember);
```
객체를 그대로 리턴할바엔 ok() 만 사용하는것도 괜찮을듯

Galmaeki replied:
api명세서 나왔으니 따라가면 좋을듯

Galmaeki reviewed (changes requested):
PR이름 컨벤션 따라 변경 바람!
제목 컨벤션 만들어놨음 #2 
라벨이랑 기여자 추가!!

추가로 해야 할 일 :
- [ ] SecurityConfig 작성
- [ ] 비밀번호 암호화
- [ ] 액세스 토큰 사용

prozen99 committed:
feat : dto 추가
SignupRequest = > loginid ,nickname, password 관리
SignupResponse = > 회원가입 응답 처리

Galmaeki commented on src/main/resources/application.yaml:
```diff
@@ -3,7 +3,7 @@ spring:
     url: jdbc:mysql://localhost:3306/calender
     username: root
-    password: ${LOCAL_DB_PASSWORD}
+    password: skhuA+4.5
```
커밋 할 때 어떤 파일 커밋 되는 지 잘 확인 해야 함!!
로컬에서 쓰는 비번이라도 퍼블릭으로 올라가선 좋을게 없음!!

Galmaeki commented on src/main/java/discordstudy/calender/domain/member/dto/SignupRequest.java:
```diff
@@ -0,0 +1,12 @@
+@Getter
+@Setter
+public class SignupRequest {
+    private String loginId;
```
DTO는 불변객체로 만드는게 권장됨
final 키워드 붙일것
or Record 클래스를 사용하여 보일러 플레이트를 줄이고 불변을 보장하는 방법이 있음
```java
public record SignupRequest(
        String loginId,
        String nickname,
        String password
) {
}
```
[레코드를 알아보자](https://colevelup.tistory.com/28)

Galmaeki commented on src/main/java/discordstudy/calender/domain/member/controller/MemberController.java:
```diff
@@ -0,0 +1,28 @@
+    {
+        Member savedMember=memberService.registermember(request);
+        SignupResponse response=new SignupResponse(savedMember.getLoginId(),savedMember.getNickname());
+        return ResponseEntity.ok(response);
```
상태코드만 리턴하는 방식은 이렇게 사용 가능
```java
    @PostMapping("/signup")
    public ResponseEntity<Void> signup(@RequestBody SignupRequest request)
    {
        //do something...
        return ResponseEntity.ok().build();
    }
```

Galmaeki commented on src/main/java/discordstudy/calender/domain/member/controller/MemberController.java:
```diff
@@ -0,0 +1,28 @@
+    @PostMapping("/signup")
+    public ResponseEntity<SignupResponse> signup(@RequestBody SignupRequest request)
+    {
+        Member savedMember=memberService.registermember(request);
```
엔티티를 Controller가 알게 할 필요 없이 바로 Response 객체를 리턴하는 방법도 있음

Galmaeki commented on src/main/java/discordstudy/calender/domain/member/dto/SignupResponse.java:
```diff
@@ -0,0 +1,16 @@
+    public SignupResponse(String loginId, String nickname) {
+        this.loginId = loginId;
+        this.nickname = nickname;
+    }
```
생성자에서 객체를 받아 바로 만드는 방법 도 존재함
이 경우는 DTO가 엔티티와 의존이 강해지는 문제가 있으나 코드가 깔끔하고 가독성이 높아짐
균형을 잘 잡는것이 중요
```java
public SignupResponse(Member member) {
    loginId = member.getLoginId;
    nickname = member.getNickname;
}
````

Galmaeki commented on src/main/java/discordstudy/calender/domain/member/service/MemberService.java:
```diff
@@ -0,0 +1,26 @@
+    public Member registermember(SignupRequest request)
+
+    {
+        Member member=new Member();
```
빈 엔티티 객체를 만들어 set로 넣는것 보다는
엔티티 생성 코드는 엔티티의 생성자에 위임하여 만드는 편이 코드를 더 깔끔하게 작성 할 수 있음
비즈니스와 관련 없는 보일러 플레이트는 가독성을 해치는 경향이 있음

prozen99 replied:
코드 캡처가 Member member=new Member() 부분만 되어있어서 그런데 밑에 코드까지 포함해서 말한건가?

Galmaeki replied:
내!

Galmaeki reviewed (changes requested):
수고햇지만 아직 갈길이 멀다!

prozen99 committed:
Refactor:
SignupRequest : final 키워드 추가 생성자 추가
SignupResponse :Member 엔티티를 직접이용해서 id와 nickname 구성
MemberController: 응답을 빈객체로 보내는 부분 구성
MemberService : entity 생성자 이용

Member: entity에 생성자 추가

Galmaeki commented on src/main/java/discordstudy/calender/domain/member/dto/SignupRequest.java:
```diff
@@ -0,0 +1,19 @@
+import lombok.Setter;
+
+@Getter
+@Setter
```
필드가 final이니 빠져도 될 것 같음

Galmaeki reviewed (approved):
고생햇슴
나머지 제안사항 적용할지 머지할지 선택해서 머지하거나 추가 커밋 하면 댈듯
